src/module-loader/dotnet/src/MorganStanley.ComposeUI.ModuleLoader/Runners/NativeModuleRunner.cs

```diff
@@ -10,19 +10,85 @@
 // or implied. See the License for the specific language governing permissions
 // and limitations under the License.
 
+using System.Diagnostics;
+using MorganStanley.ComposeUI.ModuleLoader;
+
 namespace MorganStanley.ComposeUI.ModuleLoader.Runners;
 
 internal class NativeModuleRunner : IModuleRunner
 {
     public string ModuleType => ComposeUI.ModuleLoader.ModuleType.Native;
 
-    public Task Start(StartupContext startupContext, Func<Task> pipeline)
+    public async Task Start(StartupContext startupContext, Func<Task> pipeline)
     {
-        throw new NotImplementedException();
+        if (!startupContext.ModuleInstance.Manifest.TryGetDetails(out NativeManifestDetails details))
+        {
+            throw new Exception("Unable to get native maifest details");
+        }
+
+        startupContext.AddProperty(new EnvironmentVariables(details.EnvironmentVariables));
+
+        await pipeline();
+
+        var mainProcess = new Process();
+
+        mainProcess.StartInfo.FileName = Path.GetFullPath(details.Path.ToString());
+
+        foreach (var argument in details.Arguments)
+        {
+            mainProcess.StartInfo.ArgumentList.Add(argument);
+        }
+
+        foreach (var envVar in startupContext.GetProperties<EnvironmentVariables>().SelectMany(x => x.Variables))
+        {
+            // TODO: what to do with duplicate envvars?
+            if (!mainProcess.StartInfo.EnvironmentVariables.ContainsKey(envVar.Key))
+            {
+                mainProcess.StartInfo.EnvironmentVariables.Add(envVar.Key, envVar.Value);
+            }
+        }
+
+        var unexpectedStopHandlers = startupContext.GetProperties<UnexpectedStopCallback>();
+
+        foreach (var handler in unexpectedStopHandlers)
+        {
+            mainProcess.Exited += handler.ProcessStoppedUnexpectedly;
+        }
+
+        mainProcess.Start();
+
+        var processInfo = new MainProcessInfo(mainProcess);
+        startupContext.AddProperty(processInfo);
     }
 
-    public Task Stop(IModuleInstance moduleInstance)
+    public async Task Stop(IModuleInstance moduleInstance)
     {
-        throw new NotImplementedException();
+        MainProcessInfo? mainProcessInfo;
+
+        if ((mainProcessInfo = (MainProcessInfo?) moduleInstance.GetProperties().FirstOrDefault(x => x is MainProcessInfo)) == null) { return; }
+
+        var mainProcess = mainProcessInfo.MainProcess;
+
+        // Detach unexpected stop callbacks
+        foreach (var handler in moduleInstance.GetProperties<UnexpectedStopCallback>())
+        {
+            mainProcess.Exited -= handler.ProcessStoppedUnexpectedly;
+        }
+        var killNecessary = true;
+
+        if (mainProcess.CloseMainWindow())
+        {
+            await Task.Delay(TimeSpan.FromMilliseconds(500));
```

Comment:
This delay seems arbitrary (and too short), what if the application pops up a dialog before quitting, or just takes a longer time to shut down?

Reply:
True, it is completely arbitrary, and there are also thoughts that are not scope here to support more sophisticated shutdown flows (e.g. prompts to save) - I'll gladly change it to something more appropriate, but we should discuss what should be changed, I didn't want to change the interface without discussing.